wg-api/spec-documents/message-ports.md

```diff
@@ -0,0 +1,206 @@
+
+Here's a short example of usage on the Web:
+
+```javascript=
+const {port1, port2} = new MessageChannel();
+port1.onmessage = (e) => console.log('port1 got message', e.data);
+someOtherWindow.postMessage(
+    `here's the other end of the channel`, '*', [port2]);
+
+// In the other window
+window.onmessage = (e) => {
+  const port2 = e.ports[0]
+  port2.postMessage('send a message over the channel')
+  // port1.onmessage will be triggered
+}
+```
+
+Under the hood, this is a light wrapper around Mojo pipes and the same algorithms that `window.postMessage` uses.
+
+### Node.js worker threads
+
+Node.js took inspiration from the Channel Messaging API when they designed the [communication system for worker threads](https://nodejs.org/dist/latest-v12.x/docs/api/worker_threads.html#worker_threads_class_messagechannel).
+
+### Transferables
+
+In the `postMessage` API, one of the options available when sending a message is to pass a list of _transferables_ to be passed along with the message. These objects, once transferred through `postMessage`, are no longer accessible to the sender, and ownership is transferred to the receiver. Some examples of transferable objects are `ArrayBuffer`, `ImageBitmap` and `MessagePort`.
+
+References to transferred objects can appear in the message data, and when deserialized, the message on the receiving end will contain references to those transferred objects in the appropriate place.
+
+## Proposal
+
+I propose that we add a new IPC primitive to Electron based on and interoperable with the Channel Messaging API.
+
+### New API surface
+
+#### ipcRenderer
+
+##### `ipcRenderer.postMessage(channel, message[, options])`
+
+* `channel` String
+* `message` any
+* `options` Object (optional)
+  * `transfer` any[] - a list of transferable objects to be sent with the message.
+
+This behaves similarly to `window.postMessage`, but instead of sending a message to another window, it sends the message to the main process.
+
+Also, this behaves similarly to `ipcRenderer.send`, but only takes a single argument after the channel (the message to be sent), and allows for additional options to be specified. In particular, it allows specifying a list of transferable objects to be sent along with the message.
+
+When this message is received in the main process, any `MessagePort` objects transferred will be deserialized as `ipcMain.MessagePort`.
+
+#### Class: WebContents
+
+##### `WebContents.postMessage(channel, message[, options])`
+
+* `channel` String
+* `message` any
+* `options` Object (optional)
+  * `transfer` any[] - a list of transferable objects to be sent with the message.
+
+Similar to `WebContents.send`, but allows transfering transferable objects.
+
+> [name=Samuel Attard]
+> I'm trying to figure out if there's a way for this to be non-breaking integrated into `.send` to avoid the confusion around two different ways to send IPC.  Can't see an easy way out there though.
```

Comment:
Yeah neither. The best I came up with is that any MessagePorts sent through `send` get implicitly transferred. It's not great though.

Reply:
@MarshallOfSound